### ISSUE TITLE
Switching to English mode with a simple hotkey

### ISSUE THREAD

**Issue by yookoala** [Cangjians/ibus-cangjie#87]
Chinese user rarely only type Chinese with their computer. The below situation force user to quickly switch typing Chinese and English:
1. All terminal commands are English or roman character based character.
2. Most URLs consists solely of English characters.
3. When typing a email address in a Email content.
4. When mix using English and Chinese (common in HK).
5. When entering dash from a typing context.

Currently, the only way to switch is to switch the entire input method. Which is slow and sloppy in most case. iBus would freeze 0.1 - 0.8 second (from experience) in switching input method. Typing a sentence like "Mary 你今晚 dinner 去唔去" requires switching 3 times and a painful waiting.

Also, switching from English to Chinese (Cangjie) would involve waiting the engine to boot up to speed. There a occasion where key strokes are missed because the waiting time.

### Solutions from other OS

Chinese input engine in Microsoft Windows has an "English mode". You may enable / disable the English mode by tapping the switch button in a typing context (e.g. typing in a browser's text area).

It is an "English mode" in the Chinese input method such that you don't fall out of the Chinese input method when typing English. And switching between the 2 mode is trivial and fast. This is actually a great solution to the problem.

MacOS has the same pain in the ass and provided no solution.

**Comment by bochecha** [comment]
> Currently, the only way to switch is to switch the entire input method.

And that seems like the right UX to me.

> Which is slow and sloppy in most case. iBus would freeze 0.1 - 0.8 second (from experience) in switching input method.

That would be a bug in IBus which needs to be fixed there, not worked around here.

> Also, switching from English to Chinese (Cangjie) would involve waiting the engine to boot up to speed.

If IBus Cangjie is too slow to start, then let's fix that, not work around it.

> This is actually a great solution to the problem.

No it's not: it's a workaround to avoid fixing the real issues.

> MacOS has the same pain in the ass and provided no solution.

That's because the right solution is switching to another input source if you want to type something else than Cangjie/Quick. It's really that simple: you configure an input source for each language/input method you want to type in, and you switch between them as needed.

I switch between French and English keyboard layouts all the time, I don't expect the English keyboard layout to somehow provide a French input method.

IBus Cangjie is about inputting Chinese with the Cangjie/Quick input methods. I do not want to implement other input methods in IBus Cangjie.

**Comment by yookoala** [comment]
French and English both use romance characters. You don't really need to switch to English keyboard layout to type English. CJK language is a different situation. The need to switch English and Chinese typing is frequent and real.

The need for a Chinese input method to include English typing (pass through) is real. Passing through the raw English key in to output is hardly an input method.

**Comment by bochecha** [comment]
> You don't really need to switch to English keyboard layout to type English.

You do need to switch to a French keyboard to type French, though. I can't touch type, I look at the keyboard when I type. If I'm typing on a QWERTY keyboard (as I was back when I was living in Hong Kong) then I'll use the US layout as it maps to what I see when I look at my keyboard. But to input French, I'd need to switch to the French layout. For a word like "français" I'd type all letters with the US layout, except the "ç" for which I'd switch to the French layout temporarily.

Back when I was living in Brazil, I had to switch between French and Portuguese layouts because none really covered the other language. (missing accents, etc...)

I do understand the problem, trust me. I still think the right UX is to switch input sources, not have modes inside a single input source.

> The need for a Chinese input method to include English typing (pass through) is real.

I really don't think it is. Chinese and English are two different languages. Use a dedicated input source for each. Especially if you want to type correct English: you'd really need to use an input method, because the ubiquitous US keyboard layout doesn't have all the needed characters. (punctuation marks and quotes come to mind... and did you know English has diacritics?)

And why stop at English? Why is English any more special than other languages? Because historically Hong Kong was a British colony and many people there speak English today? What about Macao then, should we also have a Portuguese mode? What about Guangdong, should we also have a French mode?

> Passing through the raw English key in to output is hardly an input method.

But it is! I mean, for you and I who understand the difference between a keyboard layout, the IBus framework, and an engine plugin, sure, we know the difference.

But for a random non-geeky person, an input method is a way to input a given language. It's what GNOME calls "input source". Having a single input source provide "modes" for multiple languages completely breaks the mental model built by the user: there is one way to switch between languages.

I mean, why would you sometimes switch with `[shift]` and other times with `[super]+[space]`? Should it be a short or long press on `[shift]`? What about `[CapsLock]`? How do you type uppercase in the English mode? How do you get English autocorrection/autocompletion in the English mode?

All those questions (and more) are just way too complex, both to design the mode switch feature, and for users who need to constantly keep in mind what's the current status of their input method/mode and how to change that status. It's an unnecessary cognitive load, which can be entirely avoided by just sticking to the simple "there is only one way to switch between input languages".

---

TL;DR: I understand that Hong Kong people very frequently type English, even single words in the middle of sentences (or even just the "D" letter). I still consider that switching input sources is a better UX than having a passthrough mode. If the switch causes problems (e.g too slow, drops the first few inputs, ...) then let's fix those issues.

**Comment by yookoala** [comment]
To be frank, I think the biggest design flaw in iBus in general is it mix up keyboard layout with input method. They are essentially different thing.

Cangjie input method is essentially mapping writing strokes to English / romance character A-Z. It is not essential for the user to use a English (US) QWERTY keyboard layout with Cangie:

* If you're using a Dvorak keyboard, all the input keys of Cangjie should be mapped with referenced to the Dvorak keyboard layout.
* If you're using standard Japanese / Spanish keyboard, the punctuation mark should follow that on the keyboard even if you type French or English.

**Comment by bochecha** [comment]
> To be frank, I think the biggest design flaw in iBus in general is it mix up keyboard layout with input method.

IBus doesn't. GNOME does. For most non-technical users they really are the same thing, though: a way to input a language.

You're thinking they are different because one of the two can be obtained directly by typing on a keyboard (what you see painted on the key is what you get on the screen).

But that's a very limited view of how to input a language, though. (coupled to the fact that you understand the technical stack underpinning the implementations)

Think about touch devices, where the on-screen keyboard is really much closer to an input method than a keyboard layout.

Think about word suggestions, autocorrection and autocompletion, which are input method features, not something coming from a keyboard layout.

Think about how word processors have been replacing `"` by `“` (and `”`) automatically since pretty much forever, because those are the proper quotation marks and they aren't on the keyboard layout. Same with adding an unbreakable space after a `:` or replacing `-` by `–`. Those are **input method features which applications have had to reimplement** because simple keyboard layouts are so limited and don't even cover correctly the languages they were designed for.

Eventually, I'm hoping keyboard layouts will die, and everything will be input methods, so that we can all enjoy those features in every language, and applications don't need to reinvent the wheel. But we're certainly not there yet.

> If you're using a Dvorak keyboard, all the input keys of Cangjie should be mapped with referenced to the Dvorak keyboard layout.

I disagree here. Cangjie has been designed with a certain layout in mind, placing the radicals on certain keys voluntarily. There even are keyboards with the Cangjie radicals painted on them. The Cangjie layout is not "the one I'm using for another language" or even "the US QWERTY one", it is a standard "手田水口廿卜". (the first 6 characters on the top row, to reuse the convention of calling layouts AZERTY/QWERTY)

My wife has 3 input sources configured right now on her keyboard: English, French and Quick. I can imagine her confusion (and the names she'd call me) if the Quick layout changed depending on what was the input source she switched from. 🙂

**Comment by yookoala** [comment]
It is of course perfect if I can click only button on my keyboard to switch input method between English and Chinese (In Windows, that key would be left SHIFT). And it would be perfect if the transition takes no overhead.

But in reality, this is hardly achievable. We cannot, and probably should not, allocate a key just to activate when our input method is installed. Also, activating / deactivating the input method would require time to load / unload the libcangjie library and ibus-cangjie process.

**Comment by yookoala** [comment]
> I disagree here. Cangjie has been designed with a certain layout in mind, placing the radicals on certain keys voluntarily. There even are keyboards with the Cangjie radicals painted on them. The Cangjie layout is not "the one I'm using for another language" or even "the US QWERTY one", it is a standard "手田水口廿卜". (the first 6 characters on the top row, to reuse the convention of calling layouts AZERTY/QWERTY)
>
> My wife has 3 input sources configured right now on her keyboard: English, French and Quick. I can imagine her confusion (and the names she'd call me) if the Quick layout changed depending on what was the input source she switched from. :slightly_smiling_face:

Because I am a long time Cangjie user, I know for a fact that Cangjie is mapping to characters. Keyboard layout is only relevant to how people getting use to it. It is not an opinion.

Every Cangjie beginner learn the keys by sub-categories:
* A and B are Sun (日) and Moon (月); C, D, E, F, G are the five elements: Metal (金), Wood (木), Water (水), Fire (火) and Earth (土). These are the "Group of Philosophical Meaning" (哲理類);
* Then H-N would be "Group of Character Form" (字型類);
* O-R would be "Group of Body Parts" (人身類);
* S-Y, except X is reserved for other use, would be "Group of Key Stroke" (筆劃類).

This is how the input coding is designed. That's why the position of Cangjie in Dvorak keyboards should follow the position of the English character.

**Comment by yookoala** [comment]
> You're thinking they are different because one of the two can be obtained directly by typing on a keyboard (what you see painted on the key is what you get on the screen).

Yes and no.

Keyboard layout mapping are still relevant for people who are unfamiliar with key positions and depend on visual cue on the keys. For instance, I can finger type English easily. But I don't remember all the locations (plus with / without shift key) the punctuation marks. I have to actually look at my keyboard to see where things are.

For I while, I have a Japanese keyboard notebook. And typing on Linux was a pain in the ass back then. On a Japanese keyboard, all punctuation locates differently than standard QWERTY. What I actually need is to have a Japanese keyboard mapping with a Chinese input method. Back then, I have to do try-and-error every time I type punctuation.

**Comment by bochecha** [comment]
> We cannot, and probably should not, allocate a key just to activate when our input method is installed.

I think you can actually do that in the GNOME Settings (maybe in Tweak Tool?)

> Also, activating / deactivating the input method would require time to load / unload the libcangjie library and ibus-cangjie process.

The process doesn't exit when you switch away from Cangjie/Quick, so the startup overhead only happens the first time. If there is still some overhead for every switch, then it probably comes mostly from IBus and should be fixed there.

But then the IBus design where every engine is a process which must be started and then communicates over DBus is a pretty bad one:

* it introduces lots of overhead;
* every key-press gets sent on the public DBus, which is a pretty bad security flaw; (this is a keylogger)
* it plays really badly with the direction we're going with Wayland and Flatpak; (mostly the security aspects)

There have been talks of replacing IBus by something else which would be more secure and show better performance. An IBus developer had even started an experiment in that direction a few years ago (https://blog.du-a.org/2013/12/04/libtextinput/). We talk about this at pretty much every GUADEC. It will take someone to do the work though, and it will take time. But it's the right thing to do, as it would fix the performance/security concerns. (and would just make for a much nicer architecture anyway)

**Comment by bochecha** [comment]
> Keyboard layout mapping are still relevant for people who are unfamiliar with key positions and depend on visual cue on the keys. For instance, I can finger type English easily. But I don't remember all the locations (plus with / without shift key) the punctuation marks. I have to actually look at my keyboard to see where things are.

You're mixing two different things: the physical keyboard with letters painted on it, and the virtual "keyboard layout" which directly prints on screen a single letter in response to a single keycode sent by pressing a single physical key.

In my previous comments when I talked about keyboard layouts being problematic, I was speaking only about the latter.

Of course there still is a physical layout (and there always will be, as long as we're using keyboards). When I talked about replacing virtual keyboard layouts by input methods, it was obvious to me that those input methods would be based on the physical layout anyway. They would only add features on top of that.

(and I can't touch type, I've been looking at my physical keyboard the entire time I typed this comment)

**Comment by mbridon** [comment]
This has now been done over on gitlab.freedesktop.org in [MR 115](https://gitlab.freedesktop.org/cangjie/ibus-cangjie/-/merge_requests/115), closing this issue.
